### PR TITLE
Migrate @storybook/core-events to TypeScript

### PR DESCRIPTION
**Pull request by dandean**
Issue: #5030 

## What I did

Migrated `@storybook/core-events` to TypeScript.

Many packages depend on `@storybook/core-events`. Before migrating those packages to TypeScript it might make sense to migrate this one first. Also, this is my first contribution to this project and doing something small to get used to your process seems like the right way to go.

## How to test

I don't believe there is anything to test in this work. Please let me know if you'd like to see anything here.

cc @kroeder

### CODE REVIEW

**Comment by dandean** [comment]
It looks like `@storybook/ui` can't find these changes. Any pointers on what's going wrong here?

**Comment by kroeder** [comment]
I will check this later today

**Comment by ndelangen** [comment]
So awesome to have you helping us @dandean! 👏

File: lib/core-events/src/index.ts [hunk 1]
```diff
@@ -1,4 +1,20 @@
...
+  APPLY_SHORTCUT: string;
+  STORY_ADDED: string;
+  FORCE_RE_RENDER: string;
+  REGISTER_SUBSCRIPTION: string;
+  STORY_RENDERED: string;
+  STORY_ERRORED: string;
+  STORY_THREW_EXCEPTION: string;
+}
+
+const events: CoreEvents = {
```
Comment (kroeder) [inline lib/core-events/src/index.ts]: Do you think we can throw away the above interface if we use `enum` ? 
Both solutions offer auto-completion in js and ts

I did 
```typescript
enum events {
  CHANNEL_CREATED = 'channelCreated',
  GET_CURRENT_STORY = 'getCurrentStory',
  SET_CURRENT_STORY = 'setCurrentStory',
  GET_STORIES = 'getStories',
  SET_STORIES = 'setStories',
  SELECT_STORY = 'selectStory',
  APPLY_SHORTCUT = 'applyShortcut',
  STORY_ADDED = 'storyAdded',
  FORCE_RE_RENDER = 'forceReRender',
  REGISTER_SUBSCRIPTION = 'registerSubscription',
  STORY_RENDERED = 'storyRendered',
  STORY_ERRORED = 'storyErrored',
  STORY_THREW_EXCEPTION = 'storyThrewException',
}

export default events;
```
and it compiled to 
```typescript
Object.defineProperty(exports, "__esModule", { value: true });
var events;
(function (events) {
    events["CHANNEL_CREATED"] = "channelCreated";
    events["GET_CURRENT_STORY"] = "getCurrentStory";
    events["SET_CURRENT_STORY"] = "setCurrentStory";
    events["GET_STORIES"] = "getStories";
    events["SET_STORIES"] = "setStories";
    events["SELECT_STORY"] = "selectStory";
    events["APPLY_SHORTCUT"] = "applyShortcut";
    events["STORY_ADDED"] = "storyAdded";
    events["FORCE_RE_RENDER"] = "forceReRender";
    events["REGISTER_SUBSCRIPTION"] = "registerSubscription";
    events["STORY_RENDERED"] = "storyRendered";
    events["STORY_ERRORED"] = "storyErrored";
    events["STORY_THREW_EXCEPTION"] = "storyThrewException";
})(events || (events = {}));
exports.default = events;
```

before it was 
```typescript
var events = {
    CHANNEL_CREATED: 'channelCreated',
    GET_CURRENT_STORY: 'getCurrentStory',
    SET_CURRENT_STORY: 'setCurrentStory',
    GET_STORIES: 'getStories',
    SET_STORIES: 'setStories',
    SELECT_STORY: 'selectStory',
    APPLY_SHORTCUT: 'applyShortcut',
    STORY_ADDED: 'storyAdded',
    FORCE_RE_RENDER: 'forceReRender',
    REGISTER_SUBSCRIPTION: 'registerSubscription',
    STORY_RENDERED: 'storyRendered',
    STORY_ERRORED: 'storyErrored',
    STORY_THREW_EXCEPTION: 'storyThrewException',
};
exports.default = events;
```

**Comment by kroeder** [comment]
You can fix the `yarn.lock` conflict by just running `yarn` locally and push the changed file

`yarn bootstrap --core` works for me locally..
@ndelangen any idea why it is failing in CI?

File: lib/core-events/src/index.ts (hunk 1 above)
Comment (dandean) [reply lib/core-events/src/index.ts]: Works for me – I'll give that change a shot.

**Comment by ndelangen** [comment]
error in CI:
`storybook/core-events`

**Comment by ndelangen** [comment]
@dandean I invited you to the GitHub organisation,
You should now be able to push to branches on this repo, no need for forking around ;)

**Comment by ndelangen** [comment]
found a FIX *I think*: the prepare script was missing in `package.json`.

**Comment by dandean** [comment]
Wonderful – thank you @ndelangen. I've got the change requests from @kroeder on my TODO list today so if this builds as expected now I'll get those patches in.

**Review by kroeder — APPROVED** [review]

**Comment by ndelangen** [comment]
Amazing work @dandean ! 👏

**Comment by dandean** [comment]
Thanks for your help, @kroeder and @ndelangen.

**Comment by tmeasday** [comment]
I'm not sure I fully understand the subtlety but I want to flag that I had to make this change: https://github.com/storybooks/storybook/commit/5cfd3c82b88f0921f8e70a7056e672223f308da2

And it looks like a lot of other files use the events in similar ways. I'm not sure what needs to change but hopefully someone that has more time and context can fill me in please?

**Comment by dandean** [comment]
@tmeasday Thanks for noting this. I think I understand this issue, and that it's a conflict between how the module systems operate.

In the CommonJS module system (depending on how babel is configured) using `module.export = { ... }`, the export is both the default export (enabling `import Events from '@storybook/core-events'`) and a destructurable object (enabling `import { CHANNEL_CREATED } from '@storybook/core-events'`).

That duality isn't automatically supported in the ES6 and TypeScript module systems, so I think we need to make it explicit by both exporting the `enum` as the default export _and_ exporting each constant as a named export.

I will put a PR together.

**Comment by dandean** [comment]
WIP patch here: https://github.com/storybooks/storybook/pull/5186

**Pull request merged**